installer/ssh_cluster.go

```diff
@@ -0,0 +1,545 @@
+}
+
+func (p privateKeySigner) Decrypt() (ssh.Signer, error) {
+	if p.key == nil {
+		pem, err := x509.DecryptPEMBlock(p.pem, []byte(p.Passphrase()))
+		if err != nil {
+			return nil, err
+		}
+		p.key, err = x509.ParsePKCS1PrivateKey(pem)
+		if err != nil {
+			return nil, err
+		}
+		p.Encrypted = false
+	}
+	return ssh.NewSignerFromKey(p.key)
+}
+
+func (p privateKeySigner) PublicKey() ssh.PublicKey {
+	return p.publicKey
+}
+
+func (p privateKeySigner) Sign(rand io.Reader, data []byte) (*ssh.Signature, error) {
+	signer, err := p.Decrypt()
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+	return signer.Sign(rand, data)
+}
+
+// sort []privateKeySigner with unencrypted keys first,
+// and encrypted keys with public keys before those without
+type decryptedFirst []privateKeySigner
+
+func (s decryptedFirst) Len() int      { return len(s) }
+func (s decryptedFirst) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s decryptedFirst) Less(i, j int) bool {
+	if !s[i].Encrypted && s[j].Encrypted {
+		return true
+	}
+	if s[i].Encrypted && !s[j].Encrypted {
+		return false
+	}
+	return s[i].publicKey != nil && s[j].publicKey == nil
+}
+
+func (c *SSHCluster) findSSHKeySigners() (signers []privateKeySigner) {
+	keyDir := filepath.Join(c.sshDir())
+	if stat, err := os.Stat(keyDir); err != nil || !stat.IsDir() {
+		return
+	}
+	walkFunc := func(path string, info os.FileInfo, err error) error {
+		if info.IsDir() {
+			return nil
+		}
+		if strings.HasSuffix(path, ".pub") {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil
+		}
+		b, _ := pem.Decode(data)
+		if b == nil {
+			return nil
+		}
+		s := privateKeySigner{
+			base:      c.base,
+			path:      path,
+			pem:       b,
+			Encrypted: x509.IsEncryptedPEMBlock(b),
+		}
+		if s.Encrypted {
+			publicKeyPath := fmt.Sprintf("%s.pub", path)
+			if stat, err := os.Stat(publicKeyPath); err == nil && !stat.IsDir() {
+				if data, err := ioutil.ReadFile(publicKeyPath); err == nil {
+					pk, _, _, _, err := ssh.ParseAuthorizedKey(data)
+					if err == nil {
+						s.publicKey = pk
+					}
+				}
+			}
+			signers = append(signers, s)
+			return nil
+		}
+		privateKey, err := x509.ParsePKCS1PrivateKey(b.Bytes)
+		if err != nil {
+			return nil
+		}
+		s.key = privateKey
+		signers = append(signers, s)
+		return nil
+	}
+	filepath.Walk(keyDir, walkFunc)
+	sort.Sort(decryptedFirst(signers))
+	return
+}
+
+func (c *SSHCluster) knownHostsPath() string {
+	return filepath.Join(c.sshDir(), "known_hosts")
+}
+
+func (c *SSHCluster) sshDir() string {
+	if runtime.GOOS == "windows" {
+		return filepath.Join(os.Getenv("USERPROFILE"), ".ssh")
+	}
+	return filepath.Join(config.HomeDir(), ".ssh")
+}
+
+func (c *SSHCluster) createKnownHosts() error {
+	var buf bytes.Buffer
+	k, err := knownhosts.Unmarshal(&buf)
+	if err != nil {
+		return err
+	}
+	c.knownHosts = k
```

Comment:
This method is weird, it seems like there should be an easier way of making a new known hosts structure? Also, this should be called `newKnownHosts`.